Descobrindo a senha
===================



<hr>

## Objetivo



Usar um algoritmo genético para descobrir uma senha.



## Descrição do problema



Neste problema, a função objetivo deve saber a senha correta e quantificar de alguma maneira o quão perto ou longe os palpites estão da solução (veja que isso é algo que não temos no mundo real. Nenhum site irá te dizer se você está acertando ou errando seu palpite). O critério de parada deste problema é quando a senha for descoberta.



> <hr>

## Importações



In [1]:
from funcoes import populacao_inicial_senha
from funcoes import funcao_objetivo_pop_senha
from funcoes import selecao_torneio_min
from funcoes import cruzamento_ponto_simples as funcao_cruzamento
from funcoes import mutacao_senha
# 'funcoes' é um arquivo com o objetivo de reunir as funções que podem ser reutilizadas em outros trabalhos
import random

## Códigos e discussão



<div style=' text-align: justify; text-justify: inter-word;'>
    Esse problema pode parecer mais difícil do que os anteriores, já que não estamos mais falando de poucas opções de número ou de número em geral, mas sim de letras. Contudo, ainda assim devemos pensar em indivíduos compostos por genes, sendo que esses genes agora corresponderão às letras que formarão as senhas.
    <br>O objetivo desse problema é encontrar uma senha textual. Não se trata mais de um problema de maximização envolvendo caixas binárias, não binárias e números. Queremos um indivíduo cujo genes corresponderão a letras e queremos diminuir a distância dessas letras em relação às letras que compõem a senha buscada. Em outras palavras, queríamos encontrar a senha sugerida com menor distância (0) em relação a senha dada, caracterizando, então, um problema de minimização. Apesar disso, entendo que fazer essa afirmação poderia até mesmo me confundir, caso eu pensasse que toda a estrutura do trabalho mudou. O problema é diferente, mas é preciso recordar que ainda se trata de algoritmos genéticos, como já pontuei também no parágrafo anterior.

In [ ]:
#### código do Professor Daniel, responsável pela disciplina 

### CONSTANTES

# relacionadas à busca ## podem ser alteradas
TAMANHO_POP = 50
#NUM_GERACOES = 2000 # nem precisa mais ser uma constante, pq não serve mais como critério de parada
CHANCE_CRUZAMENTO = 0.5
CHANCE_MUTACAO = 0.05
NUM_COMBATENTES_NO_TORNEIO = 3

# relacionadas ao problema a ser resulvido ## não podem ser alteradas
SENHA = "correcthorsebatterystaple" 
# vamos estudar, letra a letra, a distância entre as letras da senha sugerida e da real
# se mudarmos a senha real, a senha na qual queremos chegar, não vai dar certo
# não vamos conseguir minimizar nada, já que nossa repsosta final tá sempre mudando

LETRAS_POSSIVEIS = "abcdefghijklmnopqrstuvwxyz"
# em suma, todas as letras do alfabeto são possíveis

NUM_GENES = len(SENHA)
# o número genes possíveis corresponde ao número de caracteres da senha

In [7]:
# funções locais
# feitas somente para esse problema, por isso senha e letras entram como argumentos
# temos funções parciais aqui, caracterizadas por chamar outras funções 

def cria_populacao_inicial(tamanho, tamanho_senha): 
    return populacao_inicial_senha(tamanho, tamanho_senha, LETRAS_POSSIVEIS)

def funcao_objetivo_pop(populacao):
    return funcao_objetivo_pop_senha(populacao, SENHA)

def funcao_selecao(populacao, fitness):
    return selecao_torneio_min(populacao, fitness, NUM_COMBATENTES_NO_TORNEIO)

def funcao_mutacao(individuo):
    return mutacao_senha(individuo, LETRAS_POSSIVEIS)

# iniciando o problema
populacao = cria_populacao_inicial(TAMANHO_POP, NUM_GENES) 
# população é criada de acordo com o número de indivíduos definido no início do problema (TAM_POP)
# o segundo critério para criá-la é o número de genes que compõem um indivíduo

melhor_fitness_ja_visto = float("inf")  # 'inf' = infinito em python

print("Progresso da melhor senha já vista:")

#for n in range(NUM_GERACOES):  
    # isso tá errado, por isso estou deixando comentado
    # nosso critério de parada agora passa a ser quando a senha correta foi encontrada, não é mais o n° de gerações

while melhor_fitness_ja_visto != 0: # distância é zero quando descobrimos a senha
    
    # Seleção
    fitness = funcao_objetivo_pop(populacao)
    populacao = funcao_selecao(populacao, fitness)
    
    # Cruzamento
    # divide-se a população em dois grupos: pais e mães
    # para separar os indivíduos, indicamos que a escolha deve ser feita de maneira intercalada
    pais = populacao[0::2] # a partir do primeiro elemento da população, pega-se um e pula outro
    maes = populacao[1::2] # a partir do segundo elemento da população, pega-se um e pula outro
    
    contador = 0
    
    for pai, mae in zip(pais, maes): # zip é uma função interna que permite trabalhar com elementos de 'pais' e 'maes' ao mesmo tempo
        if random.random() <= CHANCE_CRUZAMENTO:
            filho1, filho2 = funcao_cruzamento(pai, mae) # o cruzamento de um pai e uma mãe origina dois filhos 
            populacao[contador] = filho1
            populacao[contador + 1] = filho2
        
        contador = contador + 2   
        
    # Mutação
    for n in range(len(populacao)):
        if random.random() <= CHANCE_MUTACAO:
            individuo = populacao[n]
            populacao[n] = funcao_mutacao(individuo)   
    # mutação vai acontecer em um individuo escolhido aleatoriamente na população
    # mas mutação só vai acontecer para valores inferiores ou iguais à chance de mutação
            
    # melhor individuo já visto até agora
    fitness = funcao_objetivo_pop(populacao)
    menor_fitness = min(fitness) # trata-se de um problema de minimização, portanto, o melhor fitness é o menor
    if menor_fitness < melhor_fitness_ja_visto:        
        posicao = fitness.index(menor_fitness)
        melhor_individuo_ja_visto = populacao[posicao] # melhor indivíduo já visto seria um hall da fama modificado, já que se quer 1 só
        melhor_fitness_ja_visto = menor_fitness # conforme valores menores de fitness forem sendo encontrados, o melhor valor muda
        print("".join(melhor_individuo_ja_visto), "- fitness:", melhor_fitness_ja_visto)

        
print()
print("Melhor palpite da senha encontrado:")
print("".join(melhor_individuo_ja_visto))
print()
print()
# os dois últimos print possuem somente o objetivo de dar um espaço maior entre o final dos resultados e a discussão

Progresso da melhor senha já vista:
fquaecieitvizewegbjputotj - fitness: 170
dkphodwsnshbecgaukwqiaced - fitness: 154
msrupqxkrjaffdullskzdcvmc - fitness: 151
dkphodwsnshbecgaukwqiavmc - fitness: 142
dkphoftkushbecgaukwqiavmc - fitness: 138
dkphodwsnshdfdullkwqiaced - fitness: 122
dkphodwsnshdfdullswqiaced - fitness: 116
dkphoftkushbecullkwqiavmc - fitness: 106
dkptoftkushdfdullswqiaced - fitness: 104
dkphoftkushdfduwlswqiavmc - fitness: 95
dkptoftkushdfduwlswqiavmc - fitness: 87
dkptoftkushdfduwlswqiavme - fitness: 85
dkptoftkusldfduwlswqiavmd - fitness: 82
dkpteftkushdfduwlswqiavme - fitness: 75
dkpteftkusldfduwlswqiavme - fitness: 71
dkpteftkusndfduwlswqiavme - fitness: 69
dkpteftkusndfduwjswqiavme - fitness: 67
dppteftkusndfduwjswqiavme - fitness: 64
dprteftkusndfduwjswqiavme - fitness: 62
dprteftkusnefduwjswqiavme - fitness: 61
dprteftkusnefduwjswqtavme - fitness: 50
dprteftkusndfduweswqtavme - fitness: 46
dprteftkpsnefduwjswqtavme - fitness: 45
dprseftkusnefduweswqtavme - fitness

<div style=' text-align: justify; text-justify: inter-word;'>
<br> Assim como tinhamos diversas opções de números entre 0 e 100 no experimento 04, temos aqui várias opções de letras disponíveis também. Para conseguir qual delas mais se aproxima do nosso objetivo, entra um cálculo que considera a distância das letras sugeridas às letras da senha real. Essa distância, então, é calculada de letra a letra.
<br> Já que o problema mudou, o código também sofreu alterações, mas nada muito gigante. O que se pode observar nas novas funções disponíveis no 'funcoes.py' é que aquelas relacionadas com: genes, indivíduos, população, função objetivo, mutação e seleção. Pode-se justificar grande parte das mudanças pelo fato de que agora estamos trabalhando com letras, senhas e distância. E também pelo fato de que buscamos resolver um problema de minimização e a roleta máxima utilizada na seleção, por exemplo, se torna inutilizável nese caso. Com isso em mente, o código adquiriu algumas particularidades. Recursos foram utilizados para possibilitar a manipulação das letras como números, já que, dessa forma, seria possível medir a distância entre as senhas sugeridas e a senha desejada. Tal recurso foi a função 'ord' utilizada dentro da função  'funcao_objetivo_senha' (presente no arquivo 'funcoes.py', também disponível na pasta 'AlgoritmosGeneticos' no repositório do GitHub). Além disso, como já mencionada anteriormente, esse problema se trata de uma questão de minimização, já que queremos encontrar a senha menos diferente da nossa constante previamente definida o possível. Para tanto, no momento de seleção dos indivíduos, o método de seleção mais adequado foi o por torneio, no qual três indivíduos eram comparados e se escolhia aquele com valor de fitness inferior aos seus combatentes. O interessante da seleção por torneio é que, nela, nenhum indivíduo (tendo ele sido selecionado ou não) morre ou é descartado, eles apenas retornam ao grupo de possíveis combatentes e um mesmo indivíduo pode ser selecionado mais de uma vez. Considerando as alterações, ao rodar o código, nota-se que depois de algumas tentativas, finalmente obtém-se o valor '0', o que demonstra que a senha sugerida é identica à senha real. 
</div>

> <hr>

## Conclusão



> <div style=' text-align: justify; text-justify: inter-word;'>
    A distância dos resultados encontrados em relação a senha buscada (dada pelo fitness) foi diminuindo progressivamente. Achei interessante que a distância da primeira senha sugerida é bem grande, maior do que eu imaginava que era possível. Contudo, é interessante também utilizar essa confusão de maneira a relembrar que o valor do fitness não representa quantas letras estão distantes da senha original ou algo do tipo. Na realidade, as letras assumem valores de números específicos para que o computador consiga entedê-las, então, o valor de fitness mostra que há diferenreça, sem implicar que 170 letras estão diferentes.</div>
    
> <div style=' text-align: justify; text-justify: inter-word;'> Considero importante ressaltar a natureza desse exercício: ele foi elaborado com o objetivo de que exercitásemos conceitos em contextos diferentes. É improvável que uma situação real desse tipo ocorra, não haveria um sistema que te informe a senha sugerida, por exemplo, então o que nos importa aqui é entender o funcionamento do código e do Algortimo Genético e não pensar em como ele seria traduzido para uma situação do cotidiano.</div>

> <div style=' text-align: justify; text-justify: inter-word;'> Antes de finalizar, gostaria de expor uma das questões que fiquei pensando bastante durante a aula. Antes de vir para a sala para tentar resolver o problema, eu dedico um tempo para fazer anotações anteriores (que posteriormente foram aderidas à discussão. Isso tem sido bastante importante para mim, porque me permite compreender o problema estudado com mais profundidade do que simplesmente já partir para o código. Anotar o que é necessário e expor quais são mudanças são necessárias e por quais razões elas devem ser feitas me ajuda a acompanhar a aula mais adequadamente e me ajuda também a compreender as discussões estabelecidas. Em suma, considero ser sempre benéfico buscar compreender o seu problema com profundida antes de já partir para a resolução prática dele.</div>

<hr>

## Playground



**Demais anotações durante a aula**
- Mutação pode retardar o processo de encontrar todas as letras corretas, mas não impede de encontrá-las. Contudo, é importante ter o número de gerações solicitadas e ter os operadores necessários;
- Tentar pensar durante as aulas em quais momentos podemos usar a característica da flexibilidade que algoritmos genéticos nos permitem;
- Em problemas de várias variáveis, a superfície da resposta é instável. Ter certeza de que os 'ruins' não vão te ajudar pode ser problemático e pode atrapalhar o desempenho.